// @workos-inc/node/lib/sso/interfaces/connection-type.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionType = void 0;
var ConnectionType;
(function (ConnectionType) {
    ConnectionType["ADFSSAML"] = "ADFSSAML";
    ConnectionType["AdpOidc"] = "AdpOidc";
    ConnectionType["Auth0SAML"] = "Auth0SAML";
    ConnectionType["AzureSAML"] = "AzureSAML";
    ConnectionType["CasSAML"] = "CasSAML";
    ConnectionType["ClassLinkSAML"] = "ClassLinkSAML";
    ConnectionType["CloudflareSAML"] = "CloudflareSAML";
    ConnectionType["CyberArkSAML"] = "CyberArkSAML";
    ConnectionType["DuoSAML"] = "DuoSAML";
    ConnectionType["GenericOIDC"] = "GenericOIDC";
    ConnectionType["GenericSAML"] = "GenericSAML";
    ConnectionType["GoogleOAuth"] = "GoogleOAuth";
    ConnectionType["GoogleSAML"] = "GoogleSAML";
    ConnectionType["JumpCloudSAML"] = "JumpCloudSAML";
    ConnectionType["KeycloakSAML"] = "KeycloakSAML";
    ConnectionType["LastPassSAML"] = "LastPassSAML";
    ConnectionType["MagicLink"] = "MagicLink";
    ConnectionType["MicrosoftOAuth"] = "MicrosoftOAuth";
    ConnectionType["MiniOrangeSAML"] = "MiniOrangeSAML";
    ConnectionType["NetIqSAML"] = "NetIqSAML";
    ConnectionType["OktaSAML"] = "OktaSAML";
    ConnectionType["OneLoginSAML"] = "OneLoginSAML";
    ConnectionType["OracleSAML"] = "OracleSAML";
    ConnectionType["PingFederateSAML"] = "PingFederateSAML";
    ConnectionType["PingOneSAML"] = "PingOneSAML";
    ConnectionType["RipplingSAML"] = "RipplingSAML";
    ConnectionType["SalesforceSAML"] = "SalesforceSAML";
    ConnectionType["ShibbolethGenericSAML"] = "ShibbolethGenericSAML";
    ConnectionType["ShibbolethSAML"] = "ShibbolethSAML";
    ConnectionType["SimpleSamlPhpSAML"] = "SimpleSamlPhpSAML";
    ConnectionType["VMwareSAML"] = "VMwareSAML";
})(ConnectionType = exports.ConnectionType || (exports.ConnectionType = {}));
